### PR TITLE
[feat]: [CDS-40785]: Rollback pipeline on CG

### PR DESCRIPTION
**Pull request by lasdox**
## Describe your changes

This feature creates a option to rollback the entire pipeline on failure/ approval rejection.

This is behind a feature flag: SPG_PIPELINE_ROLLBACK

## How it works?

We created a new EnvState, the EnvRollbackState, which is nothing more than a wrapper responsible to trigger the rollback execution.
When this FF is enabled, we update the StateMachine with the corresponding EnvRollbackStates. The StateMachine is also updated with a new rollbackPipelineMap, which will map the failure state to the corresponding EnvRollbackState that should be triggered.

- When a stage fails, it will call the PipelineStageExecutionAdvisor for advising.

The PipelineStageExecutionAdvisor makes some checks to see if this stage is elegible to the rollbackPreviousStages feature.
If its true, it will create an advisor stating the next state name as the EnvRollbackState that should be triggering the rollback

- StateMachineExecutor is responsible to continue the execution with the new advise
Here the executor will find the nextState, clone the StateExecutionInstance, set all the properties and continue the execution.
It is responsible to filling all the data that the EnvRollbackState needs.

- EnvRollbackState execution
This is just another step, it is responsible to finding the workflowExecutionId that needs to be rolledback. Then this will just trigger and send the executor back the data to show the UI.

## What if a stage already has a rollback failure strategy?
The StateMachineExecutor is responsible of checking if the WorkflowExecution already has rollbackInstances. If this is true, it will skip this execution and go to the next EnvRollbackState.

## Limitations
The only limitation know so far is with multi service workflows.
When a multi service workflow has no previous executions in one of his infras, the rollback will fail for all infras.
This is because we are using the onDemand rollback behind the scenes, and it does not handle well single States with multiple infras.

## Comment Triggers
<details>
  <summary>Build triggers</summary>
  
- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>
  
  You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`
  
- Compile: `trigger compile`
- CodeFormat: `trigger codeformat`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- runDockerizationCheck: `trigger dockerizationcheck`
- runAuthorCheck: `trigger authorcheck`
- Checkstyle: `trigger checkstyle`
- PMD: `trigger pmd`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/36885)

### CODE REVIEW

**Comment by diptiman-adak** [comment]
Quality gate Quality gate: **_Failed_** 
 


            
 #### 6 Issues 
* Code issue 5 Code issues 
* Anti-pattern 1 Anti-patterns 

 [Details](https://app.embold.io/organization/gh/harness/repositories/7752dbb3099d2dba3b5d37f5931a5e4c/pull-request/36885)

**Review by hextechpal — APPROVED** [review]
> Reviewed 1 of 6 files at r1, 2 of 8 files at r2, 20 of 20 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/36885)* status: Some of the checks are too old.

**Comment by lasdox** [comment]
trigger ti0

**Comment by lasdox** [comment]
trigger ti1

**Comment by lasdox** [comment]
trigger ti2

**Comment by lasdox** [comment]
trigger ti4

**Comment by lasdox** [comment]
trigger ti4

**Comment by lasdox** [comment]
trigger ft2

**Comment by lasdox** [comment]
trigger ti0

**Review by hextechpal — APPROVED** [review]
> Reviewed 29 of 30 files at r4, 1 of 1 files at r5, 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/36885)* status: :shipit: complete! ✓

File: 400-rest/src/main/java/software/wings/beans/Pipeline.java [hunk 1]
```diff
@@ -102,6 +102,7 @@ public static List<MongoIndex> mongoIndexes() {
   public static final String NAME_KEY = "name";
   public static final String DESCRIPTION_KEY = "description";
 
+  public boolean rollbackPreviousStages;
```
Comment (abhinav-harness) [inline 400-rest/src/main/java/software/wings/beans/Pipeline.java:105]: make this Boolean

**Review by abhinav-harness — APPROVED** [review]
> Reviewed 1 of 6 files at r1, 1 of 8 files at r2, 6 of 20 files at r3, 29 of 30 files at r4, 1 of 1 files at r5, 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/36885)* status: :shipit: complete! ✓

**Comment by lasdox** [comment]
trigger codebasehash

**Pull request merged**